Cancellation for Executable

gerhardol opened the pull request:
Fixes #9440
#9440 is mostly closed in #9505, this handles a few effects
Closes #9501

## Proposed changes

Long running process like "git difftool --tool-help" are currently never cancelled, they must run to exit.
I have seen this command require over a minute at work.
After #9505, there is no major problem due to this, it is possible to open the console and close GE quickly after opening, but it could be a potential issue.

- CancellationToken for Executable
Possibility to cancel processes, not hanging until closing.
There may be other ways to do this too.

- CancellationToken handling for CustomDiffMergeTools
Runs in the background and was never cancelled before. This add cancellation token handling and also cancels the Git command.
A few other cleanups, review commit by commit

## Test methodology <!-- How did you ensure quality? -->

Manual

* Start Settings to clear the cache for merge/diff tools
* F12 for command log
* Open `Solve merge conflicts...` form
* Notice `mergetool --tool-help` in the list
* Close ResolveConflicts
* Notice that the time is updated immediately, no exit code is written (before the command always ran to completion

## Merge strategy

<!-- Change the following if the merge strategy should be changed:
- Squash merge (maintainer to decide merge message, PR submitter should cleanup commits/messages at PR approval).
- Rebase merge (PR submitter must change the commit message for the last commit).
- Merge commit. (PR submitter to rebase and squash before merges).
- To be decided later.
The maintainer may still request the contributor to squash and rebase, to make sure that merges and commit messages are clarified.
-->

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

drewnoakes commented on GitCommands/Git/Executable.cs:
```diff
@@ -232,6 +233,9 @@ public StreamReader StandardError
             public Task<int> WaitForExitAsync() => _exitTaskCompletionSource.Task;
 
+            /// <inheritdoc />
+            public Task WaitForExitAsync(CancellationToken token) => _process.WaitForExitAsync(token);
```
Can this return the exit code, for parity with sibling methods?

Also, why is this method implemented so differently to the overload that does not support cancellation?

gerhardol replied:
It is likely possible to get exit status etc from the returned task, but exceptions must be handled in the same way the exitTask handles it now.

To be able to cancel the process, the process must be involved. The other interfaces are add-on to return data from the process.
This method is really something different than the overload without cancellation. Maybe the existing method should be renamed to `WaitForExitTaskAsync` to clarify this.

Maybe there are entirely better ways to do this?

drewnoakes commented on Plugins/GitUIPluginInterfaces/IProcess.cs:
```diff
@@ -55,10 +56,17 @@ public interface IProcess : IDisposable
+        /// </summary>
+        /// <param name="token">An optional token to cancel the asynchronous operation.</param>
+        /// <returns>A task that will complete when the process has exited, cancellation has been requested, or an error occurs.</returns>
+        Task WaitForExitAsync(CancellationToken token);
```
Is cancellation expected to kill the process, or just stop the wait? It's not clear from the docs here.

gerhardol replied:
Doc copied from the library doc. I can clarify.
Cancellation throws, the statement after this never executes.

gerhardol replied:
I believe it is killed - I cannot see the process running after cancellation and `_process.Exited` is invoked

gerhardol commented on GitCommands/CustomDiffMergeToolCache.cs:
```diff
@@ -59,20 +56,31 @@ public async Task<IEnumerable<string>> GetToolsAsync(GitModule module, int delay
             // The command will compete with other resources, avoid delaying startup
             // Do not bother with cancel tokens
-            await Task.Delay(delay);
+            await Task.Delay(delay, cancellationToken);
+
+            await _mutex.WaitAsync(cancellationToken).ConfigureAwait(false);
```
Not needed, `WaitAsync(cancellationToken)` will throw
Unclear from documentation, but in the debugger the second line is never executed

The implementation mostly follows this pattern already, just two occurrences changed in 176ef8aec30c88773104478558b8fe723aec8908

gerhardol commented on GitCommands/CustomDiffMergeToolCache.cs:
```diff
@@ -59,20 +56,31 @@ public async Task<IEnumerable<string>> GetToolsAsync(GitModule module, int delay
+                if (_tools is null)
+                {
+                    // Parsing has failed, just provide an empty list, no user notification
+                    _tools = new List<string>();
```
Not accepted by the compiler , the correct System.Array is referenced
```
Error	CS0426	The type name 'Empty<>' does not exist in the type 'Array'	GitCommands	C:\dev\gc\gitextensions_4\GitCommands\CustomDiffMergeToolCache.cs	81	Active
```

RussKie replied:
Is correct namespace referenced? https://docs.microsoft.com/dotnet/api/system.array.empty

RussKie replied:
I pulled down your branch, and it compiles for me
![image](https://user-images.githubusercontent.com/4403806/130347284-b0a3b5e7-2f21-43ba-859f-15200523e3c1.png)

gerhardol replied:
Now it worked, no clue why it could not be compiled before

> I pulled down your branch, and it compiles for me

This not changed in this PR (compare to the separate commit with cleanup I added..)

gerhardol commented on GitCommands/Submodules/SubmoduleStatusProvider.cs:
```diff
@@ -293,8 +293,8 @@ private static string GetModuleBranch(string path, string noBranchText)
         {
             _previousSubmoduleUpdateTime = DateTime.Now;
-            cancelToken.ThrowIfCancellationRequested();
             await TaskScheduler.Default;
+            cancelToken.ThrowIfCancellationRequested();
```
Not sure

RussKie replied:
Why was it necessary to change this?
![image](https://user-images.githubusercontent.com/4403806/130347316-1987832c-32f4-4bf7-87b8-50074ba5c3be.png)

gerhardol replied:
It is not necessary, it is a separate commit with cleanups to follow the normal patterns

gerhardol commented on GitUI/BranchTreePanel/RepoObjectsTree.Nodes.Submodules.cs:
```diff
@@ -188,7 +188,6 @@ internal async Task SetStatusToolTipAsync(CancellationToken token)
                 }
 
                 await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(token);
-                token.ThrowIfCancellationRequested();
```
See https://github.com/gitextensions/gitextensions/pull/9522#discussion_r693329169

gerhardol commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -1590,9 +1590,12 @@ private void OnShowSettingsClick(object sender, EventArgs e)
-            RevisionGrid.LoadCustomDifftools();
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await new CustomDiffMergeToolProvider().ClearAsync();
+                revisionDiff.LoadCustomDifftools();
+                RevisionGrid.LoadCustomDifftools();
```
LoadCustomDifftools() has cancellations in each form.

Clear would benefit from cancellations, now Clear will first wait for the previous loading to complete.
This is not be a major problem, really just 0:10 to maybe 1:30 after start or Settings last run (and starting merge conflicts), up to twice the time to see changes in the drop down list is not a major difference.
So a revDiff/revGrid (as well as merge conflicts) `CancelLoadCustomDiffTools` is not making much difference
Maybe a comment

```
// Await previous loading of tools to finish (no cancellation) and clear the existing lists
```

gerhardol replied:
Edit: Changed to separate async

gerhardol commented on GitUI/CustomDiffMergeToolProvider.cs:
```diff
@@ -18,10 +20,11 @@ public class CustomDiffMergeToolProvider
         {
-            CustomDiffMergeToolCache.DiffToolCache.Clear();
-            CustomDiffMergeToolCache.MergeToolCache.Clear();
+            await CustomDiffMergeToolCache.DiffToolCache.ClearAsync();
+            await CustomDiffMergeToolCache.MergeToolCache.ClearAsync();
```
In this routine it makes no difference at all as there is virtually no execution time in the methods themselves.
What (could) take time is waiting for the previous custom loading to finish.

So in FormBrowse the load of diff tool do not have to wait on cancel of mergetools.

Will switch the order so conflict merges do not have to await difftools.

gerhardol replied:
Changed this, see https://github.com/gitextensions/gitextensions/pull/9522#discussion_r693333350

gerhardol commented:
> We need to kill the process when cancellation is requested

See https://github.com/gitextensions/gitextensions/pull/9522#discussion_r693329169

RussKie commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -1590,9 +1590,17 @@ private void OnShowSettingsClick(object sender, EventArgs e)
-            revisionDiff.LoadCustomDifftools();
-            RevisionGrid.LoadCustomDifftools();
+
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await new CustomDiffMergeToolProvider().ClearAsync(isDiff: false);
+            }).FileAndForget();
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await new CustomDiffMergeToolProvider().ClearAsync(isDiff: true);
+                revisionDiff.LoadCustomDifftools();
+                RevisionGrid.LoadCustomDifftools();
+            }).FileAndForget();
```
This deserves at least a comment why we can't combine these two.

gerhardol replied:
added comment

gerhardol commented:
Minor review comments, rebased to add cancellation after #9524 was merged https://github.com/gitextensions/gitextensions/pull/9524#pullrequestreview-735475144

RussKie reviewed (approved):
👍

gerhardol commented:
Rebased and squashed and updated commit messages, rebase merge tomorrow

Renamed one more exitTask method
![image](https://user-images.githubusercontent.com/6248932/130357478-d0e45eba-4f18-4643-afec-16156876c2a1.png)

RussKie commented on GitCommands/Git/Executable.cs:
```diff
@@ -230,12 +231,15 @@ public StreamReader StandardError
+            public Task WaitForExitAsync(CancellationToken token) => _process.WaitForExitAsync(token);
+
+            /// <inheritdoc />
+            public int WaitForExitTask()
```
This name is confusing and wrong. It returns an `int` and not a `task`. The same goes for the async version.

gerhardol replied:
The method "waits for the exitTask" created when the process exits, including the extra work to add to log etc.
It is not a normal `WaitForExit()` that waits for the process task to exit.

I still think the name is better.
But what about `WaitForProcessExit()`?

drewnoakes replied:
It is the non-async form of `WaitForExitAsync` so should be called `WaitForExit` in order to be consistent with the rest of the framework (e.g. `Read`/`ReadAsync`, `Write`/`WriteAsycn`, ...).

gerhardol replied:
But there are two variants of WaitForExit() now: One that waits for the process to exit (that I added, cancellable) and the existing that waits for the additional exitTask to exit. I find the new method to follow the intention of the WaitForExit() method better, but I can rename that instead.

drewnoakes commented on GitCommands/Git/Executable.cs:
```diff
@@ -230,12 +231,15 @@ public StreamReader StandardError
 
             /// <inheritdoc />
-            public Task<int> WaitForExitAsync() => _exitTaskCompletionSource.Task;
+            public Task<int> WaitForExitTaskAsync() => _exitTaskCompletionSource.Task;
```
Similarly this should be `WaitForExitAsync`. Async methods don't generally contain the word 'task'.

gerhardol replied:
Async or not is the same.
The name is for the exitTask.

RussKie reviewed (approved):
👍 

Once on .NET 6.0 we will have this: https://devblogs.microsoft.com/dotnet/new-dotnet-6-apis-driven-by-the-developer-community/#waitasync-improvements

gerhardol commented:
> 👍

Was that for calling the existing exitTask as `WaitForProcessExit()` or reverting to `WaitForExit()` and rename the new cancellable method (call that `WaitForProcessExit()` instead?`)?

> Once on .NET 6.0 we will have this: https://devblogs.microsoft.com/dotnet/new-dotnet-6-apis-driven-by-the-developer-community/#waitasync-improvements

Unfortunately no change here, the additional exitTask is the problem here it seems.

RussKie commented:
It's almost a midnight here, and you're asking complex questions :D

RussKie commented on Plugins/GitUIPluginInterfaces/IProcess.cs:
```diff
@@ -47,18 +48,25 @@ public interface IProcess : IDisposable
+        /// </summary>
+        /// <param name="token">An optional token to cancel the asynchronous operation.</param>
+        /// <returns>A task that will complete when the process has exited, cancellation has been requested, or an error occurs.</returns>
+        Task WaitForExitAsync(CancellationToken token);
```
So now we have:
```cs
/// Blocks the calling thread until the process exits, or when this object is disposed.
int WaitForProcessExit();

/// Returns a task that completes when the process exits, or when this object is disposed.
Task<int> WaitForProcessExitAsync();

/// Instructs the process component to wait for the associated process to exit, or for the cancellationToken to be cancelled (cancells the process).
Task WaitForExitAsync(CancellationToken token);
```

Maybe we can do something like this, and remove the need for the new method?
```cs
            public Task<int> WaitForProcessExitAsync(CancellationToken token)
            {
                token.Register(() => _exitTaskCompletionSource.SetCanceled(token));
                return _exitTaskCompletionSource.Task;
            }
```

gerhardol replied:
Not really an option, exitTask is awaiten together with tasks for stdout and possibly stderr.

gerhardol replied:
I did not get this working, just renaming

gerhardol commented:
Rebased, renamed `WaitForExitAsync` back and the new cancellation method to `WaitForProcessExitAsync` (was not able to merge them, probable possible but lets settle with this).

RussKie commented:
We can iterate on this later, if necessary.

gerhardol commented:
> We can iterate on this later, if necessary.

I am considering setting the priority for the long running and background processes to BelowNormal
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.priorityclass?view=net-5.0

Probably very little difference for users though

RussKie commented:
How do we tell whether a process is expected to be a long running one?

On Tue, 31 Aug 2021 at 16:39, Gerhard Olsson ***@***.***>
wrote:

> We can iterate on this later, if necessary.
>
> I am considering setting the priority for the long running and background
> processes to BelowNormal
>
> https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.priorityclass?view=net-5.0
>
> Probably very little difference for users though
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9522#issuecomment-908944236>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXUATB43HA6XOQ4MXS3T7R2LXANCNFSM5CPDMNEA>
> .
>

gerhardol commented:
> How do we tell whether a process is expected to be a long running one?

By knowledge, those we know may be long (currently these two in the PR).
In the same way GE could decrease prio for `git --no-optional-locks` and maybe no console visible.

But this fit in other PRs. I may look at this after the blame/filter PRs are merged
